libraries/package-deps-hash/src/getRepoState.ts

```diff
@@ -246,6 +246,109 @@ export function applyWorkingTreeState(
   }
 }
 
+/**
+ * Get the list of all git submodule paths
+ * @param rootDirectory - The root directory of the Git repository
+ * @param gitPath - The path to the Git executable
+ * @internal
+ */
+export function getSubmodulePaths(rootDirectory: string, gitPath?: string): string[] {
+  const submodulePaths: string[] = [];
+
+  const submoduleStatusResult: child_process.SpawnSyncReturns<string> = Executable.spawnSync(
+    gitPath || 'git',
+    ['submodule', 'status'],
+    {
+      currentWorkingDirectory: rootDirectory
+    }
+  );
+
+  if (submoduleStatusResult.status !== 0) {
+    ensureGitMinimumVersion(gitPath);
+
+    throw new Error(
+      `git submodule status exited with status ${submoduleStatusResult.status}: ${submoduleStatusResult.stderr}`
+    );
+  }
+
+  // <hash> <path> (heads/main)\n
+  submoduleStatusResult.stdout.split('\n').forEach((line) => {
+    if (line) {
+      const [, submodulePath] = line.trim().split(' ');
+      submodulePaths.push(submodulePath);
+    }
+  });
+  return submodulePaths;
+}
+
+/**
+ * Augments the state value with modifications that are in git submodules
+ * @param rootDirectory - The root directory of the Git repository
+ * @param state - The current map of git path -> object hash. Will be mutated.
+ * @param gitPath - The path to the Git executable
+ * @internal
+ */
+export function applySubmodulesTreeState(
+  rootDirectory: string,
+  state: Map<string, string>,
+  gitPath?: string
+): void {
+  const submodulePaths: string[] = getSubmodulePaths(rootDirectory, gitPath);
+
+  if (submodulePaths.length === 0) {
+    return;
+  }
+
+  const submoduleLsTreeResult: child_process.SpawnSyncReturns<string> = Executable.spawnSync(
+    gitPath || 'git',
+    ['submodule', 'foreach', 'git', '--no-optional-locks', 'ls-tree', '-r', '--full-name', 'HEAD', '--'],
+    {
+      currentWorkingDirectory: rootDirectory
+    }
+  );
+
+  if (submoduleLsTreeResult.status !== 0) {
+    ensureGitMinimumVersion(gitPath);
+
+    throw new Error(
+      `git ls-tree for submodules exited with status ${submoduleLsTreeResult.status}: ${submoduleLsTreeResult.stderr}`
+    );
+  }
+
+  const stateBySubmodulePath: Record<string, Record<string, string>> = {};
+
+  let currentSubmodulePath: string = '';
+
+  // Entering '<submoduleFolder>'\n
+  // <mode> <type> <newhash>\t<path>\n
+  submoduleLsTreeResult.stdout.split('\n').forEach((line) => {
+    if (line) {
+      line = line.trim();
+      const submodulePath: string | undefined = line.match(/'(.*)'/)?.[1];
+      if (submodulePath) {
+        currentSubmodulePath = submodulePath;
+        stateBySubmodulePath[currentSubmodulePath] = {};
+      } else {
+        const tabIndex: number = line.indexOf('\t');
+        const filePath: string = line.slice(tabIndex + 1);
+
+        // The newHash will be all zeros if the file is deleted, or a hash if it exists
+        const hash: string = line.slice(tabIndex - 40, tabIndex);
```

Comment:
What happens if the Git output does not correspond to the expected format (length 40)?  Ideally it should throw an exception rather than building a hash string from an incorrect character range.  The validation doesn't need to be very strict, just a sanity check.

Reply:
The logic comes from the `parseGitLsTree` at https://github.com/microsoft/rushstack/blob/ba3e7268c5472c08bb12083269413a7960598a13/libraries/package-deps-hash/src/getRepoState.ts#L41